frontend/src/metabase/nav/components/SearchBar.unit.spec.tsx

```diff
@@ -0,0 +1,182 @@
+import { Route } from "react-router";
+import userEvent from "@testing-library/user-event";
+import { waitFor, renderWithProviders, screen, within } from "__support__/ui";
+import {
+  setupRecentViewsEndpoints,
+  setupSearchEndpoints,
+} from "__support__/server-mocks";
+import {
+  createMockCollectionItem,
+  createMockModelObject,
+  createMockRecentItem,
+} from "metabase-types/api/mocks";
+import {
+  createMockSettingsState,
+  createMockState,
+} from "metabase-types/store/mocks";
+import { CollectionItem, RecentItem } from "metabase-types/api";
+import SearchBar from "metabase/nav/components/SearchBar";
+
+const TEST_SEARCH_RESULTS: CollectionItem[] = [
+  "Card ABC",
+  "Card BCD",
+  "Card DEF",
+  "Card EFG",
+].map((name, index) =>
+  createMockCollectionItem({
+    name,
+    id: index + 1,
+    getUrl: () => "/",
+  }),
+);
+
+const TEST_RECENT_VIEWS_RESULTS: RecentItem[] = [
+  "Recents ABC",
+  "Recents BCD",
+  "Recents CDE",
+  "Recents DEF",
+].map((name, index) =>
+  createMockRecentItem({
+    model_object: createMockModelObject({ name }),
+    model_id: index,
+  }),
+);
+
+const setup = ({
+  initialRoute = "/",
+  searchResultItems = TEST_SEARCH_RESULTS,
+  recentViewsItems = TEST_RECENT_VIEWS_RESULTS,
+} = {}) => {
+  const state = createMockState({
+    settings: createMockSettingsState({
+      "search-typeahead-enabled": true,
+    }),
+  });
+
+  setupSearchEndpoints(searchResultItems);
+  setupRecentViewsEndpoints(recentViewsItems);
+
+  const { history } = renderWithProviders(
+    <Route path="*" component={SearchBar} />,
+    {
+      withRouter: true,
+      initialRoute,
+      storeInitialState: state,
+    },
+  );
+
+  return { history };
+};
+
+const getSearchBar = () => {
+  return screen.getByPlaceholderText("Search…");
+};
+
+describe("SearchBar", () => {
+  describe("typing a search query", () => {
+    it("should change URL when user types a query and hits `Enter`", async () => {
+      const { history } = setup();
+
+      userEvent.type(getSearchBar(), "BC{enter}");
+
+      const location = history?.getCurrentLocation();
+      expect(location?.pathname).toEqual("search");
```

Comment:
nit: I think the reason you need to use `?` stems from `const location = history?.getCurrentLocation();`. We know for sure in this case `history` is going to be present. Can we type it somehow so that we don't need to use optional chaining when accessing `location`'s properties?

It's quite misleading as at glance, readers might think `location` could be null when it's not.

I think you should be able to force `history` to not be null in the `setup` function.

Reply:
I saw this as well - it looks like we do this a lot in the tests for `ActionCreator`, `ModelDetailPage, and a few other components. I don't know how much we can do in terms of typing since the History type seems really convoluted: 
```
(History<DefaultQuery<string>, DefaultQueryLike> & MemoryHistory & HistoryBasename & HistoryQueries) | undefined 
```

and I haven't been able to typecast to `History` without needing an `unknown` conversion.

The one thing that's worked is typing `history` as `{getCurrentLocation: () => Location}` but I feel like that adds more confusion about what `history` actually is imo

Reply:
You should be able to use `checkNotNull`

https://github.com/metabase/metabase/blob/24f1cc91069341d18650483b6e9fc366c9dc19f1/frontend/src/metabase/core/utils/types/types.ts#L11-L18

like this in the `setup` function. Then history won't be `null` or `undefined` anymore and we could get rid of `?`

```ts
  return { history: checkNotNull(history) };
```

Reply:
That's the one 😄